### PR TITLE
Allow getAttachments to return only non-nested attachments

### PR DESCRIPTION
**Pull request by couteau**
PR for changes requested in #270

### CODE REVIEW

File: tests/AttachmentTest.php [hunk 1]
```diff
@@ -42,6 +42,36 @@ public function testSaveEachAttachment()
...
+
+        // Old version took a boolean - false should be equivalent of NESTED (i.e., all attachments except inline)
+        $attachments = $Parser->getAttachments(false);
+        $this->assertCount(6, $attachments);
+
+        $attachments = $Parser->getAttachments(Parser::GA_TOPLEVEL);
+        $this->assertCount(1, $attachments);
+
+        $attachments = $Parser->getAttachments(Parser::GA_INCLUDE_INLINE);
+        $this->assertCount(2, $attachments);
```
Comment (sanmai) [inline tests/AttachmentTest.php]: FWIW I really wish there was a method for each type of attachments, to abstract away these silly arguments.

(I'm not the boss here, so take this comment with a grain of salt.)

Comment (eXorus) [reply tests/AttachmentTest.php]: I understand your point, and I don't know the best way to manage that.

For a user point of view, they want to get all attachments or sometimes only the top level attachments, or all inline attachments...
So having one function for each use case is impossible, I only see a bitmask method to fix that.

getAttachments(INLINE & TOP-LEVEL)
getAttachments()
getAttachments(INLINE)
getAttachments(TOP-LEVEL)
...

Otherwise we will have to build all the matrix:
getInlineAttachments
getTopLevelAndInlineAttachments
getTopLevelAttachments
...

or maybe both ways: bitmask method to manage all the edge cases and a few methods for the main use case.

Comment (sanmai) [reply tests/AttachmentTest.php]: Well, this `getAttachments` method, and I think I saw another, are out of place here because it appears to me handling peculiarities of loading and saving attachments is beyond responsibilities of this class. This also can be seen from the fact that these methods cannot be unit-tested, without integration tests requiring actual file operations. And the things will become even more complex if there will be need to apply the same selection logic we see here to the save attachment operation.

Therefore, as I see it, the parser should only provide a lazy-loaded list of attachments, and there should be other classes that would feed on this list, selecting and/or saving these or those attachments according to whatever logic is needed. Since these classes will receive an iterator with objects of certain interface, they could be made thoroughly unit-tested.

Comment (sanmai) [reply tests/AttachmentTest.php]: Overall I think even though this PR will add on technical debt, there's nothing unmanageable here, provided BC breaks are acceptable here, and considering there's a practical problem behind, this PR is fairly acceptable.

**Comment by eXorus** [comment]
@sanmai and @couteau  I think I'm going to merge this PR soon.

I will change later a few things.
- Remove the boolean value to do a BC
- All the const should go to another class, instead of Parser::GA_INCLUDE_INLINE, we should have AttachmentType::INCLUDE_INLINE
- We should impact also saveAttachments.
- Maybe also move the existing const for ATTACHMENT_DUPLICATE_* to a dedicated class
- Maybe switch the value from integer to binary notation
https://stitcher.io/blog/bitwise-booleans-in-php
https://stitcher.io/blog/enums-without-enums

I'm planning a new bug release 7.0 for 1 million of installations. So it's time to BC.
https://packagist.org/packages/php-mime-mail-parser/php-mime-mail-parser/stats
